compose/recomposehighlighter/src/main/java/com/example/android/compose/recomposehighlighter/RecomposeHighlighter.kt

```diff
@@ -16,100 +16,143 @@
+import androidx.compose.ui.node.DrawModifierNode
+import androidx.compose.ui.node.ModifierNodeElement
+import androidx.compose.ui.node.invalidateDraw
+import androidx.compose.ui.platform.InspectorInfo
 import androidx.compose.ui.platform.debugInspectorInfo
 import androidx.compose.ui.unit.dp
+import java.util.Objects
 import kotlin.math.min
+import kotlinx.coroutines.Job
 import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
 
 /**
  * A [Modifier] that draws a border around elements that are recomposing. The border increases in
  * size and interpolates from red to green as more recompositions occur before a timeout.
  */
 @Stable
-fun Modifier.recomposeHighlighter(): Modifier = this.then(recomposeModifier)
-
-// Use a single instance + @Stable to ensure that recompositions can enable skipping optimizations
-// Modifier.composed will still remember unique data per call site.
-private val recomposeModifier =
-    Modifier.composed(inspectorInfo = debugInspectorInfo { name = "recomposeHighlighter" }) {
-        // The total number of compositions that have occurred. We're not using a State<> here be
-        // able to read/write the value without invalidating (which would cause infinite
-        // recomposition).
-        val totalCompositions = remember { arrayOf(0L) }
-        totalCompositions[0]++
-
-        // The value of totalCompositions at the last timeout.
-        val totalCompositionsAtLastTimeout = remember { mutableStateOf(0L) }
-
-        // Start the timeout, and reset everytime there's a recomposition. (Using totalCompositions
-        // as the key is really just to cause the timer to restart every composition).
-        LaunchedEffect(totalCompositions[0]) {
+fun Modifier.recomposeHighlighter(): Modifier = this.then(RecomposeHighlighterElement())
+
+class RecomposeHighlighterElement : ModifierNodeElement<RecomposeHighlighterModifier>() {
+
+    override fun InspectorInfo.inspectableProperties() {
+        debugInspectorInfo { name = "recomposeHighlighter" }
+    }
+
+    override fun create(): RecomposeHighlighterModifier = RecomposeHighlighterModifier()
+
+    override fun update(node: RecomposeHighlighterModifier) {
+        node.incrementCompositions()
+    }
+
+    // It's never equal, so that every recomposition triggers the update function.
+    override fun equals(other: Any?): Boolean = false
+
+    override fun hashCode(): Int = Objects.hash(this)
+}
+
+class RecomposeHighlighterModifier : Modifier.Node(), DrawModifierNode {
+
+    private var timerJob: Job? = null
+
+    /**
+     * The total number of compositions that have occurred.
+     */
+    private var totalCompositions: Long = 0
+        set(value) {
+            if (field == value) return
+            restartTimer()
+            field = value
+            invalidateDraw()
+        }
+
+    /**
+     * The value of totalCompositions at the last timeout.
+     */
+    private var totalCompositionsAtLastTimeout = 0L
```

Comment:
nit: To simplify, I don't think you even need this anymore because we don't do anything with the total overtime. I think in the previous version they were using it as a way to trigger redraw, but in your version you have a better option in `invalidateDraw()`. You could just reset `totalCompositions` to `0` on the end of the timer and this would all work.